### PR TITLE
add fanout to tpu-client-next

### PR DESCRIPTION
**Pull request by KirillLykov**
#### Problem

Crate tpu-client-next doesn't have fan out so far.

#### Summary of Changes

Add this feature.

### CODE REVIEW

File: tpu-client-next/tests/connection_workers_scheduler_test.rs [hunk 1]
```diff
@@ -49,6 +50,7 @@ fn test_config(validator_identity: Option<Keypair>) -> ConnectionWorkersSchedule
         worker_channel_size: 2,
         max_reconnect_attempts: 4,
         lookahead_slots: 1,
+        leaders_fanout: Selected(1),
```
Comment (alessandrod) [inline tpu-client-next/tests/connection_workers_scheduler_test.rs]: nit: use LeadersFanout::Selected here - always use Enum::Variant

File: tpu-client-next/src/connection_workers_scheduler.rs [hunk 2]
```diff
@@ -211,3 +226,19 @@ impl ConnectionWorkersScheduler {
         WorkerInfo::new(txs_sender, handle, cancel)
     }
 }
+
+/// Splits the input vector of leaders into two parts based on the `fanout` configuration:
+/// * the first vector contains the leaders to which transactions will be sent.
+/// * the second vector contains the remaining leaders, used to warm up connections.
+fn split_leaders<'a>(
```
Comment (alessandrod) [inline tpu-client-next/src/connection_workers_scheduler.rs]: no need to return vecs here, you can do

```
let count = match {
	... => leaders.len(),
	... => count.min(leaders.len())
};
leaders.split_at(count)
```

File: tpu-client-next/src/connection_workers_scheduler.rs [hunk 3]
```diff
@@ -115,39 +128,41 @@ impl ConnectionWorkersScheduler {
...
+                        max_reconnect_attempts,
+                    );
+                    workers.push(*new_leader, worker).await;
                 }
-            };
 
-            // Regardless of who is leader, add future leaders to the cache to
-            // hide the latency of opening the connection.
+                tokio::select! {
+                    send_res = workers.send_transactions_to_address(new_leader, transaction_batch.clone()) => match send_res {
```
Comment (alessandrod) [inline tpu-client-next/src/connection_workers_scheduler.rs]: Right now this will block if the corresponding worker's channel is full, so
effectively the slowest leader will slow down all other leaders.

I think we should 1) increase the channel size and 2) add a
try_send_transactions_to_address that returns an error if the channel is full.
If the channel is full, I guess as a start we could log a warning and drop the
batch for that leader, but at least we don't slow down the other leaders.

Then longer term, we need to think of a better API so that the caller of the
crate can decide what should happen: drop the batch? increase the channel size?
slow down upstream?

File: tpu-client-next/src/connection_workers_scheduler.rs [hunk 4]
```diff
@@ -39,6 +39,15 @@ pub enum ConnectionWorkersSchedulerError {
     LeaderReceiverDropped,
 }
 
+/// This enum defines to how many discovered leaders we will send transactions.
+pub enum LeadersFanout {
+    /// Send transactions to all the leaders discovered by the `next_leaders`
+    /// call.
+    All,
+    /// Send transactions to the first selected number of leaders.
+    Selected(usize),
```
Comment (alessandrod) [inline tpu-client-next/src/connection_workers_scheduler.rs]: I'd call this `Next(usize)` so it's like `LeadersFanout::Next(4)` which sounds
better to me.

The doc should also clarify that the value is clamped by the number of leaders
returned by next_leaders()

File: tpu-client-next/src/leader_updater.rs [hunk 5]
```diff
@@ -35,7 +35,7 @@ pub trait LeaderUpdater: Send {
     /// If the current leader estimation is incorrect and transactions are sent to
     /// only one estimated leader, there is a risk of losing all the transactions,
     /// depending on the forwarding policy.
-    fn next_leaders(&self, lookahead_slots: u64) -> Vec<SocketAddr>;
+    fn next_leaders(&mut self, lookahead_slots: u64) -> Vec<SocketAddr>;
```
Comment (alessandrod) [inline tpu-client-next/src/leader_updater.rs]: why this change?

**Review by alessandrod — CHANGES_REQUESTED** [review]

File: tpu-client-next/src/connection_workers_scheduler.rs (hunk 3 above)
Comment (KirillLykov) [reply tpu-client-next/src/connection_workers_scheduler.rs]: This makes sense, I'm not sure that there is a value in having both `send_transactions` and `try_send_transactions` in the context of this crate. So probably just modify the code to use `try_`. This also allows to reduce `select!`

Comment (KirillLykov) [reply tpu-client-next/src/connection_workers_scheduler.rs]: The size of the channel `worker_channel_size` is configurable through the Scheduler config. The only question what value should be used in practice for the STS, for example.

File: tpu-client-next/src/leader_updater.rs (hunk 5 above)
Comment (KirillLykov) [reply tpu-client-next/src/leader_updater.rs]: Because to implement this trait in STS I need to able to modify self.

**Comment by KirillLykov** [comment]
@alessandrod all addressed except for 0rtt (next time)

File: tpu-client-next/tests/connection_workers_scheduler_test.rs [hunk 6]
```diff
@@ -46,9 +48,10 @@ fn test_config(validator_identity: Option<Keypair>) -> ConnectionWorkersSchedule
         stake_identity: validator_identity,
         num_connections: 1,
         skip_check_transaction_age: false,
-        worker_channel_size: 2,
+        worker_channel_size: 100,
```
Comment (KirillLykov) [inline tpu-client-next/tests/connection_workers_scheduler_test.rs:55]: Due to the change from `send_transactions` to `try_send_transactions`, had to increase it to avoid surprises in CI.

File: tpu-client-next/src/connection_workers_scheduler.rs [hunk 7]
```diff
@@ -202,12 +225,26 @@ impl ConnectionWorkersScheduler {
...
     }
 }
+
+/// Splits the input vector of leaders into two parts based on the `fanout` configuration:
+/// * the first vector contains the leaders to which transactions will be sent.
+/// * the second vector contains the remaining leaders, used to warm up connections.
+fn split_leaders<'a>(
+    leaders: &'a [SocketAddr],
+    fanout: &'a LeadersFanout,
+) -> (&'a [SocketAddr], &'a [SocketAddr]) {
```
Comment (illia-bobyr) [inline tpu-client-next/src/connection_workers_scheduler.rs]: *minor*

It is actually better to use distinct lifetimes here.
I think, calling a function that combines two potentially unrelated lifetimes likes this could cause problems in the caller code.

In particular, this function establishes a constraint that the returned values need to live as long as the value referenced by the `fanout` argument.  Which is not actually the case.
But the caller can not know that, by just looking at the function type.

```suggestion
fn split_leaders<'leaders, 'fanout>(
    leaders: &'leaders [SocketAddr],
    fanout: &'fanout LeadersFanout,
) -> (&'leaders [SocketAddr], &'leaders [SocketAddr]) {
```

Also, I think, it is better to give lifetimes proper names.
Just like it is better to name variables.
Even considering that the lifetime names scope is mostly short.

File: tpu-client-next/src/send_transaction_stats.rs [hunk 8]
```diff
@@ -164,3 +164,47 @@ impl fmt::Display for SendTransactionStats {
...
+            }
+        }
+    };
+}
+
+// Define the non-atomic struct and the `to_non_atomic` conversion method
+define_non_atomic_struct!(
+    SendTransactionStatsNonAtomic,
+    SendTransactionStats,
+    {
```
Comment (illia-bobyr) [inline tpu-client-next/src/send_transaction_stats.rs]: *minor*

I would call the macro `define_non_atoic_version_for!`.  Just to be extra clear, that it creates a struct based on another struct.

```suggestion
// Define the non-atomic struct and the `to_non_atomic` conversion method
define_non_atomic_version_for!(
    SendTransactionStats,
    SendTransactionStatsNonAtomic,
    {
```

File: tpu-client-next/src/workers_cache.rs [hunk 9]
```diff
@@ -37,35 +33,30 @@ impl WorkerInfo {
...
-        txs_batch: TransactionBatch,
-    ) -> Result<(), WorkersCacheError> {
-        self.sender
-            .send(txs_batch)
-            .await
-            .map_err(|_| WorkersCacheError::ReceiverDropped)?;
+    fn try_send_transactions(&self, txs_batch: TransactionBatch) -> Result<(), WorkersCacheError> {
+        self.sender.try_send(txs_batch).map_err(|err| match err {
+            mpsc::error::TrySendError::Full(_) => WorkersCacheError::FullChannel,
+            mpsc::error::TrySendError::Closed(_) => WorkersCacheError::ReceiverDropped,
```
Comment (illia-bobyr) [inline tpu-client-next/src/workers_cache.rs]: *style*
Consider importing `mpsc::error::TrySendError`.

```suggestion
use {
    /* ... */
    tokio::{sync::mpsc::{self, mpsc::error::TrySendError}, task::JoinHandle},
    /* ... */
};
    /* ... */

            TrySendError::Full(_) => WorkersCacheError::FullChannel,
            TrySendError::Closed(_) => WorkersCacheError::ReceiverDropped,
```

File: tpu-client-next/src/workers_cache.rs [hunk 10]
```diff
@@ -86,99 +80,113 @@ pub enum WorkersCacheError {
...
+            });
+        }
+        None
     }
 
     /// Sends a batch of transactions to the worker for a given peer. If the
     /// worker for the peer is disconnected or fails, it is removed from the
     /// cache.
-    pub async fn send_transactions_to_address(
+    pub(crate) fn try_send_transactions_to_address(
```
Comment (illia-bobyr) [inline tpu-client-next/src/workers_cache.rs:124]: *minor*

Unrelated to this PR, but the `address` suffix in the name is probably unnecessary.
As the call will have the address as the first argument, it is probably readable enough if the function is called `try_send_transactions()` or `try_send_transactions_to()`.

File: tpu-client-next/src/workers_cache.rs [hunk 11]
```diff
@@ -86,99 +80,113 @@ pub enum WorkersCacheError {
...
+
+    pub(crate) async fn shutdown(self) -> Result<(), WorkersCacheError> {
+        self.worker.shutdown().await
+    }
+}
+
+pub(crate) fn maybe_shutdown_worker(worker: Option<ShutdownWorker>) {
+    if let Some(worker) = worker {
+        tokio::spawn(async move {
+            let leader = worker.leader();
```
Comment (illia-bobyr) [inline tpu-client-next/src/workers_cache.rs]: *style*

You can reduce the whole body indentation with a `let/else`:
```suggestion
    let Some(worker) = worker else {
        return;
    };

    tokio::spawn(async move {
        let leader = worker.leader();
```

File: tpu-client-next/src/workers_cache.rs [hunk 12]
```diff
@@ -86,99 +80,113 @@ pub enum WorkersCacheError {
...
+    }
+
+    pub(crate) async fn shutdown(self) -> Result<(), WorkersCacheError> {
+        self.worker.shutdown().await
+    }
+}
+
+pub(crate) fn maybe_shutdown_worker(worker: Option<ShutdownWorker>) {
+    if let Some(worker) = worker {
+        tokio::spawn(async move {
```
Comment (illia-bobyr) [inline tpu-client-next/src/workers_cache.rs]: Spawning tasks without observing their execution, I think, is a bit fragile.

The executor will keep track of them, but it is not clear in the rest of the code as to what is going on.
And if those tasks take a long time to complete or even hang, it would be just reflected in the executor taking a long time to finish or hanging.

I do not fully understand the end to end logic here.
So, it could be hard to track.
But I suggest you consider putting all the shutdown tasks in an `UnorderedFutures` or something like that, allowing you to track and even interrupt the shutdown process.

In particular, it would allow the shutdown errors to be processed in a centralized location, rather than them being just printed as warnings.
Though again, not sure if you really need this feature.
But, maybe for collection of the stats?

File: tpu-client-next/tests/connection_workers_scheduler_test.rs (hunk 6 above)
Comment (illia-bobyr) [reply tpu-client-next/tests/connection_workers_scheduler_test.rs:55]: Maybe add a comment here in the code?
You have collected some knowledge about this configuration parameter.
But, unless you write it down in a comment, nobody else will be able to benefit from it.
No one will find this PR comment after the PR is closed.

File: tpu-client-next/tests/connection_workers_scheduler_test.rs [hunk 13]
```diff
@@ -433,10 +436,10 @@ async fn test_staked_connection() {
 
     // Wait for the exchange to finish.
     tx_sender_shutdown.await;
-    let localhost_stats = join_scheduler(scheduler_handle).await;
+    let localhost_stats = join_scheduler(scheduler_handle).await.to_non_atomic();
```
Comment (illia-bobyr) [inline tpu-client-next/tests/connection_workers_scheduler_test.rs]: Would it make sense to put the `.to_non_atomic()` call inside the `join_scheduler()`?
Or we do not want it there for non-test case?

Comment (KirillLykov) [reply tpu-client-next/tests/connection_workers_scheduler_test.rs]: You are right, it allows to avoid cloning by doing this.

File: tpu-client-next/src/connection_workers_scheduler.rs (hunk 3 above)
Comment (KirillLykov) [reply tpu-client-next/src/connection_workers_scheduler.rs]: one thing that worries me is the scenario with `transaction-bench`. There we actually need to have a backpressure (which has been created with these `worker_channel.send().await`) because it tries to generate as many as possible transaction batches and we rely on slowing down sending them.
So I want to add configuration flag controlling option `try_send` or `send`

Comment (KirillLykov) [reply tpu-client-next/src/connection_workers_scheduler.rs]: In case of STS, `try_send` is fine because even if we drop some transactions silently, there is a retry thread which sends them again anyways.

File: tpu-client-next/src/workers_cache.rs (hunk 12 above)
Comment (alessandrod) [reply tpu-client-next/src/workers_cache.rs]: > But I suggest you consider putting all the shutdown tasks in an
> UnorderedFutures or something like that, allowing you to track and even
> interrupt the shutdown process.

we explicitly don't want to do this tho: we don't want cleanup to delay sending
transactions. So we do need tasks. We could have a JoinSet and call
poll_join_next to poll without blocking and pop tasks off the join set, although
that seems extra work for little gain. Background tasks don't block the runtime https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#shutdown

File: tpu-client-next/src/workers_cache.rs (hunk 10 above)
Comment (KirillLykov) [reply tpu-client-next/src/workers_cache.rs:124]: This I will not touch in this PR to minimize the number of changes (to simplify the backport 2.1)

File: tpu-client-next/src/connection_workers_scheduler.rs [hunk 14]
```diff
@@ -115,49 +131,55 @@ impl ConnectionWorkersScheduler {
...
-                    break;
                 }
-            };
+            }
 
-            // Regardless of who is leader, add future leaders to the cache to
-            // hide the latency of opening the connection.
+            // add future leaders to the cache to hide the latency of opening the
+            // connection.
             for peer in future_leaders {
```
Comment (KirillLykov) [inline tpu-client-next/src/connection_workers_scheduler.rs]: Here it established connection to future leaders

File: tpu-client-next/src/connection_workers_scheduler.rs [hunk 15]
```diff
@@ -40,14 +40,27 @@ pub enum ConnectionWorkersSchedulerError {
...
+
 /// This enum defines to how many discovered leaders we will send transactions.
 pub enum LeadersFanout {
     /// Send transactions to all the leaders discovered by the `next_leaders`
     /// call.
     All,
     /// Send transactions to the first selected number of leaders discovered by
     /// the `next_leaders` call.
-    Next(usize),
+    Next(Fanout),
```
Comment (alessandrod) [inline tpu-client-next/src/connection_workers_scheduler.rs]: You don't like `Next { send: usize, connect: usize}`?

File: tpu-client-next/src/connection_workers_scheduler.rs [hunk 16]
```diff
@@ -114,50 +133,49 @@ impl ConnectionWorkersScheduler {
...
                         max_reconnect_attempts,
+                        stats.clone(),
                     );
-                    workers.push(*peer, worker).await;
+                    maybe_shutdown_worker(workers.push(*peer, worker));
+                }
+            }
+
+            for new_leader in fanout_leaders {
+                if !workers.contains(new_leader) {
```
Comment (alessandrod) [inline tpu-client-next/src/connection_workers_scheduler.rs:159]: nit: I don't think that this can ever happen? I'd remove the code

**Review by alessandrod — APPROVED** [review]
approving with a nit. Feel free to fix the nit in one of the followups if you don't want to do another CI run

File: tpu-client-next/src/connection_workers_scheduler.rs [hunk 17]
```diff
@@ -202,12 +221,31 @@ impl ConnectionWorkersScheduler {
...
         });
 
         WorkerInfo::new(txs_sender, handle, cancel)
     }
 }
+
+/// Splits `leaders` into two slices based on the `fanout` configuration:
+/// * the first slice contains the leaders to which transactions will be sent,
+/// * the second vector contains the leaders, used to warm up connections. This
+///   slice includes the the first set.
```
Comment (illia-bobyr) [inline tpu-client-next/src/connection_workers_scheduler.rs:237]: *minor*

```suggestion
/// * the second slice contains the leaders, used to warm up connections. This
///   slice includes the first set.
```

File: tpu-client-next/src/connection_workers_scheduler.rs [hunk 18]
```diff
@@ -114,50 +133,49 @@ impl ConnectionWorkersScheduler {
...
+            for peer in connect_leaders {
                 if !workers.contains(peer) {
+                    let stats = send_stats_per_addr.entry(peer.ip()).or_default();
                     let worker = Self::spawn_worker(
                         &endpoint,
                         peer,
                         worker_channel_size,
                         skip_check_transaction_age,
                         max_reconnect_attempts,
+                        stats.clone(),
```
Comment (illia-bobyr) [inline tpu-client-next/src/connection_workers_scheduler.rs:152]: *minor*

This `clone()` should be unnecessary - I do not see `stats` used in this block anymore.

```suggestion
                        stats,
```

File: tpu-client-next/src/connection_workers_scheduler.rs [hunk 19]
```diff
@@ -114,50 +133,49 @@ impl ConnectionWorkersScheduler {
...
-                    break;
-                }
-            };
+            let updated_leaders = leader_updater.next_leaders(leaders_fanout.connect);
 
-            // Regardless of who is leader, add future leaders to the cache to
-            // hide the latency of opening the connection.
-            for peer in future_leaders {
+            let (fanout_leaders, connect_leaders) =
+                split_leaders(&updated_leaders, &leaders_fanout);
```
Comment (illia-bobyr) [inline tpu-client-next/src/connection_workers_scheduler.rs:140]: It is a bit confusing that we call it `send` and `connect` portions elsewhere, but here they are called `fanout` and `connection` portions.

Maybe it would be more consistent to call it `send_leaders` and `connect_leaders` here as well?

```suggestion
            let (send_leaders, connect_leaders) =
                split_leaders(&updated_leaders, &leaders_fanout);
```

File: tpu-client-next/src/connection_workers_scheduler.rs (hunk 16 above)
Comment (illia-bobyr) [inline tpu-client-next/src/connection_workers_scheduler.rs:159]: Similar to the `fanout` to `send` rename above, the `new_leader` name is a bit confusing to me here.
I would expect the `new_leader` name to imply we are going to open a connection to this leader or start a worker for it.

But we actually start workers in the block above.

Maybe call it `send_to` instead?
Or some other name that indicates that this is only a destination for the next transaction batch.
It could as well be the same leader as in the previous slot group[^1].

```suggestion
            for send_to in send_leaders {
                if !workers.contains(send_to) {
```

[^1]: Is there a name for a sequence of `NUM_CONSECUTIVE_LEADER_SLOTS` slots?

Comment (illia-bobyr) [reply tpu-client-next/src/connection_workers_scheduler.rs:159]: This can happen if a connection to the leader is dropped and the worker is stopped.
This arm in the `send_err` match case below:
```rust
                    Err(WorkersCacheError::ReceiverDropped) => {
                        // Remove the worker from the cache, if the peer has disconnected.
                        maybe_shutdown_worker(workers.pop(*new_leader));
                    }
```

It is possible for the `fanout_leaders` to contain duplicates.
The duplicate would not be able to get a matching worker.

File: tpu-client-next/src/connection_workers_scheduler.rs (hunk 18 above)
Comment (KirillLykov) [reply tpu-client-next/src/connection_workers_scheduler.rs:152]: weird that it passed clippy

**Comment by KirillLykov** [comment]
@ilya-bobyr yeah, in the follow up these renamings and also need to think how to have a backpressure for sending transactions. Not obvious to me so far, but this backpressure is not needed for SendTransactionService which is priority for now, only for transaction-bench.
